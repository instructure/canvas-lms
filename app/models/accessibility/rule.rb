# frozen_string_literal: true

#
# Copyright (C) 2025 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

module Accessibility
  class Rule
    class << self
      attr_accessor :id, :link

      def registry
        rules = [
          Accessibility::Rules::AdjacentLinksRule,
          Accessibility::Rules::HeadingsSequenceRule,
          Accessibility::Rules::HeadingsStartAtH2Rule,
          Accessibility::Rules::ImgAltFilenameRule,
          Accessibility::Rules::ImgAltLengthRule,
          Accessibility::Rules::ImgAltRule,
          Accessibility::Rules::LargeTextContrastRule,
          Accessibility::Rules::ListStructureRule,
          Accessibility::Rules::ParagraphsForHeadingsRule,
          Accessibility::Rules::SmallTextContrastRule,
          Accessibility::Rules::TableCaptionRule,
          Accessibility::Rules::TableHeaderRule,
          Accessibility::Rules::TableHeaderScopeRule
        ]

        rules.index_by(&:id)
      end

      def pdf_registry
        [
          Accessibility::Rules::HasLangEntryRule,
        ]
      end

      # Tests if an element passes this accessibility rule
      # @param elem [Nokogiri::XML::Element] The element to test
      # @return [String] nil if the element passes, the validation problem
      # if there's an issue
      def test(elem)
        raise NotImplementedError, "#{self} must implement/override test"
      end

      # Gets a form definition for correcting the issue
      # @param _elem [Nokogiri::XML::Element] The element in the content to fix
      # @return [FormField] Form field for correcting the issue
      def form(_elem)
        {}
      end

      # Generates a fix for the issue in the content, to be used in the
      # correction form
      # @param _elem [Nokogiri::XML::Element] The element in the content to fix
      # @return [String] The generated fix, or nil if generation is not supported
      # @throws [StandardError] if the fix cannot be generated for a reason
      def generate_fix(_elem)
        nil
      end

      # Fixes the issue in content, changing it in-place
      # @param _elem [Nokogiri::XML::Element] The element in the content to fix
      # @param _value [String] The value received back from the correction form
      # @throws [StandardError] if the fix cannot be applied for a reason
      # @throws [NotImplementedError] if not implemented in a rule
      # @return [String] Potential error message if the fix fails, or nil if it
      # succeeds
      def fix!(_elem, _value)
        raise NotImplementedError, "#{self} must implement fix"
      end

      # Gets the name of this rule
      # @return [String] The name of this rule
      def display_name
        raise NotImplementedError, "#{self} must implement display_name"
      end

      # Gets the message for users about this rule
      # @return [String] The message about this issue
      def message
        raise NotImplementedError, "#{self} must implement message"
      end

      # Gets the explanation of why this issue is important
      # @return [String] The explanation of the issue
      def why
        raise NotImplementedError, "#{self} must implement/override why"
      end
    end
  end
end
