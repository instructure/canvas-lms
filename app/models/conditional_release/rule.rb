# frozen_string_literal: true

#
# Copyright (C) 2020 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

module ConditionalRelease
  class Rule < ActiveRecord::Base
    include Deletion

    validates :course_id, presence: true
    validates :trigger_assignment_id, presence: true
    validate :trigger_assignment_in_same_course
    belongs_to :trigger_assignment, :class_name => "Assignment"

    belongs_to :course
    belongs_to :root_account, :class_name => "Account"
    has_many :scoring_ranges, -> { active.order(position: :asc) }, inverse_of: :rule, dependent: :destroy
    has_many :assignment_sets, -> { active }, through: :scoring_ranges
    has_many :assignment_set_associations, -> { active.order(position: :asc) }, through: :scoring_ranges
    accepts_nested_attributes_for :scoring_ranges, allow_destroy: true

    after_save :clear_caches

    before_create :set_root_account_id
    def set_root_account_id
      self.root_account_id ||= course.root_account_id
    end

    def trigger_assignment_in_same_course
      if trigger_assignment_id_changed? && trigger_assignment&.context_id != course_id
        errors.add(:trigger_assignment_id, "invalid trigger assignment")
      end
    end

    scope :with_assignments, -> do
      having_assignments = joins(Rule.preload_associations).group(Arel.sql("conditional_release_rules.id"))
      preload(Rule.preload_associations).where(id: having_assignments.pluck(:id))
    end

    def self.preload_associations
      { scoring_ranges: { assignment_sets: :assignment_set_associations } }
    end

    def self.includes_for_json
      {
        scoring_ranges: {
          include: {
            assignment_sets: {
              include: {assignment_set_associations: {except: [:root_account_id, :deleted_at]}},
              except: [:root_account_id, :deleted_at]
            }
          },
          except: [:root_account_id, :deleted_at]
        }
      }
    end

    def assignment_sets_for_score(score)
      AssignmentSet.active.where(scoring_range: scoring_ranges.for_score(score))
    end

    def clear_caches
      self.class.connection.after_transaction_commit do
        self.trigger_assignment.clear_cache_key(:conditional_release)
        self.course.clear_cache_key(:conditional_release)
      end
    end

    def self.is_trigger_assignment?(assignment)
      # i'm only using the cache key currently for this one case but i figure it can be extended to handle caching around all rule data fetching
      RequestCache.cache('conditional_release_is_trigger', assignment) do
        Rails.cache.fetch_with_batched_keys('conditional_release_is_trigger', batch_object: assignment, batched_keys: :conditional_release) do
          assignment.shard.activate { self.active.where(:trigger_assignment_id => assignment).exists? }
        end
      end
    end
  end
end
