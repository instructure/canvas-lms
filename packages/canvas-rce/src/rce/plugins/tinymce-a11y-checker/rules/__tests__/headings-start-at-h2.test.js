/*
 * Copyright (C) 2023 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import rule from '../headings-start-at-h2'

describe('test', () => {
  describe('when the rule is disabled', () => {
    const config = {
      disableHeadingsStartAtH2: true,
    }

    test('returns true if the tag is not h1', () => {
      const element = document.createElement('span')
      expect(rule.test(element, config)).toBeTruthy()
    })

    test('returns true even if the tag is h1', () => {
      const element = document.createElement('h1')
      expect(rule.test(element, config)).toBeTruthy()
    })
  })

  describe('when the rule is explicitly enabled', () => {
    const config = {
      disableHeadingsStartAtH2: false,
    }

    test('returns true if the tag is not h1', () => {
      const element = document.createElement('span')
      expect(rule.test(element, config)).toBeTruthy()
    })

    test('returns false if the tag is h1', () => {
      const element = document.createElement('h1')
      expect(rule.test(element, config)).toBeFalsy()
    })
  })

  describe('when the rule is implictly enabled', () => {
    test('returns true if the tag is not h1', () => {
      const element = document.createElement('span')
      expect(rule.test(element)).toBeTruthy()
    })

    test('returns false if the tag is h1', () => {
      const element = document.createElement('h1')
      expect(rule.test(element)).toBeFalsy()
    })
  })
})

describe('data', () => {
  test("default action is 'nothing'", () => {
    expect(rule.data().action).toBe('nothing')
  })
})

describe('form', () => {
  test('returns the proper object', () => {
    expect(rule.form()).toMatchSnapshot()
  })
})

describe('update', () => {
  let body, h1

  beforeEach(() => {
    body = document.createElement('body')
    h1 = document.createElement('h1')
    body.appendChild(h1)
  })

  test('returns same element when no data is passed', () => {
    expect(rule.update(h1, {})).toBe(h1)
  })

  test('returns same element when data is missing an action', () => {
    expect(rule.update(h1, {notAction: 'something'})).toBe(h1)
  })

  test('returns same element when action is nothing', () => {
    expect(rule.update(h1, {action: 'nothing'})).toBe(h1)
  })

  test('returns h2 tag when action is elem-only', () => {
    expect(rule.update(h1, {action: 'elem-only'}).tagName).toBe('H2')
  })

  test('returns p tag when action is modify', () => {
    expect(rule.update(h1, {action: 'modify'}).tagName).toBe('P')
  })
})

describe('message', () => {
  test('returns the proper message', () => {
    expect(rule.message()).toMatchSnapshot()
  })
})

describe('why', () => {
  test('returns the proper why text', () => {
    expect(rule.why()).toMatchSnapshot()
  })
})

describe('link', () => {
  test('returns the proper link', () => {
    expect(rule.link).toMatchSnapshot()
  })
})

describe('linkText', () => {
  test('returns the proper link text', () => {
    expect(rule.linkText()).toMatchSnapshot()
  })
})
