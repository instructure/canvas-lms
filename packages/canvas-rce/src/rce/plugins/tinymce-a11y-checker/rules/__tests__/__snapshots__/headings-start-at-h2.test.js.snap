// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form returns the proper object 1`] = `
Array [
  Object {
    "dataKey": "action",
    "label": "Action to take:",
    "options": Array [
      Array [
        "nothing",
        "Leave as is",
      ],
      Array [
        "elem-only",
        "Change only this heading's level",
      ],
      Array [
        "modify",
        "Remove heading style",
      ],
    ],
  },
  Object {
    "alert": true,
    "dataKey": "alert",
    "label": "Additional considerations",
    "message": "You may need to adjust additional headings to maintain page hierarchy.",
    "variant": "warning",
  },
]
`;

exports[`link returns the proper link 1`] = `"https://www.w3.org/TR/WCAG20-TECHS/H42.html"`;

exports[`linkText returns the proper link text 1`] = `"Learn more about proper page heading structure"`;

exports[`message returns the proper message 1`] = `"The first heading on a page should be an H2."`;

exports[`why returns the proper why text 1`] = `"Webpages should only have a single H1, which is automatically used by the page's Title. The first heading in your content should be an H2."`;
